implementation/src/main/java/io/smallrye/jwt/auth/principal/KeyLocationResolver.java

```diff
@@ -52,185 +52,226 @@
  * This implements the MP-JWT 1.1 mp.jwt.verify.publickey.location config property resolution logic
  */
 public class KeyLocationResolver implements VerificationKeyResolver {
-    private static final Logger log = Logger.getLogger(KeyLocationResolver.class);
+    private static final Logger LOGGER = Logger.getLogger(KeyLocationResolver.class);
+    private static final String FAILED_TO_READ_LOCATION_MESSAGE = "Failed to read location as any of JWK, JWKS, PEM. ";
 
     // The 'content' and 'httpsJwks' fields are used to keep the key content and mutually exclusive.
     // 'content' represents the key(s) loaded from all resources but the HTTPS URL based JWK set.
-    private String content;
+    private final Supplier<String> contentSupplier;
     // 'httpsJwks' represents the JWK set loaded from the HTTPS URL.
-    private HttpsJwks httpsJwks;
+    private final Supplier<HttpsJwks> httpsJwksSupplier;
+    private final JWTAuthContextInfo authContextInfo;
 
     // If the 'smallrye.jwt.token.kid' is set then the verification key will be calculated
     // only once and used for all the token verification requests
-    private volatile PublicKey verificationKey;
+    volatile PublicKey verificationKey;
 
-    private JWTAuthContextInfo authContextInfo;
-
-    public KeyLocationResolver(JWTAuthContextInfo authContextInfo) throws UnresolvableKeyException {
+    public KeyLocationResolver(JWTAuthContextInfo authContextInfo) {
         this.authContextInfo = authContextInfo;
-        try {
-            loadContents();
-        } catch (Exception e) {
-            throw new UnresolvableKeyException("Failed to load a key from: " + authContextInfo.getPublicKeyLocation(), e);
-        }
+        this.httpsJwksSupplier = new HttpsJwksSupplier(this.authContextInfo);
+        this.contentSupplier = new ContentSupplier(this.authContextInfo);
     }
 
     @Override
     public Key resolveKey(JsonWebSignature jws, List<JsonWebStructure> nestingContext) throws UnresolvableKeyException {
-        if (verificationKey != null) {
-            return verificationKey;
-        }
-        PublicKey key = tryAsJWK(jws);
-        if (key == null) {
-            key = tryAsPEM();
-        }
-        if (key == null) {
-            throw new UnresolvableKeyException("Failed to read location as any of JWK, JWKS, PEM. "
-                    + authContextInfo.getPublicKeyLocation());
+
+        if (jws == null)
+            throw new UnresolvableKeyException(FAILED_TO_READ_LOCATION_MESSAGE);
+
+        // httpsJwks may refresh itself even if the jwksRefreshInterval is set to 0 as it checks HTTPS Cache-Control header
+        // so the public key will have to be created per request to ensure the rotation (if any) is effective.
+        String kid = getKid(jws);
```

Comment:
Instead please pass only `jws` to both `getHttpsJwk` and `getLocalJwk` and use `getKid` from those 2 methods because at this level the `getKid` call is irrelevant for the `tryAsPem` code branch

Reply:
ok